src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs

```diff
@@ -206,24 +260,105 @@ public static LogicalLocation CreateLogicalLocation(LogicalLocationVersionOne v1
         {
-            List<Notification> notifications = null;
+            var logicalLocation = new LogicalLocation
+            {
+                DecoratedName = decoratedName,
+                FullyQualifiedName = fullyQualifiedLogicalName,
+                Name = GetLogicalLocationName(fullyQualifiedLogicalName),
+                ParentKey = parentKey
+            };
+
+            return logicalLocation;
+        }
 
-            if (v1Notifications != null)
+        internal string AddLogicalLocation(LogicalLocation logicalLocation)
+        {
+            if (_currentRun.LogicalLocations == null)
             {
-                notifications = new List<Notification>();
+                _currentRun.LogicalLocations = new Dictionary<string, LogicalLocation>();
+            }
+
+            string fullyQualifiedName = logicalLocation.FullyQualifiedName;
+            string logicalLocationKey = logicalLocation.FullyQualifiedName;
+            int disambiguator = 0;
 
-                foreach (NotificationVersionOne notification in v1Notifications)
+            // We need to do the comparison here too
+            while (_currentRun.LogicalLocations.ContainsKey(logicalLocationKey))
+            {
+                LogicalLocation logLoc = _currentRun.LogicalLocations[logicalLocationKey].DeepClone();
+                logLoc.Name = GetLogicalLocationName(fullyQualifiedName);
+                logLoc.FullyQualifiedName = logicalLocation.FullyQualifiedName;
+
+                if (logicalLocation.ValueEquals(logLoc))
                 {
-                    notifications.Add(CreateNotification(notification));
+                    break;
                 }
+
+                logicalLocationKey = logicalLocation.FullyQualifiedName + "-" + disambiguator.ToString(CultureInfo.InvariantCulture);
+                disambiguator++;
+            }
+
+            if (!_currentRun.LogicalLocations.ContainsKey(logicalLocationKey))
+            {
+                _currentRun.LogicalLocations.Add(logicalLocationKey, logicalLocation);
+                FixUpLogicalLocation(logicalLocationKey);
+            }
+
+            return logicalLocationKey;
+        }
+
+        internal string GetLogicalLocationName(string fullyQualifiedLogicalName)
+        {
+            if (string.IsNullOrWhiteSpace(fullyQualifiedLogicalName))
+            {
+                throw new ArgumentNullException(nameof(fullyQualifiedLogicalName));
+            }
+
+            return fullyQualifiedLogicalName.Split(Utilities.FullyQualifiedNameDelimiters,
+                                                   StringSplitOptions.RemoveEmptyEntries).Last();
+        }
+
+        internal void FixUpAllLogicalLocations()
+        {
+            foreach (string key in _currentRun.LogicalLocations.Keys)
+            {
+                FixUpLogicalLocation(key);
+            }
+        }
+
+        internal void FixUpLogicalLocation(string key)
```

Comment:
Consider: replacing/renaming the "FixUp" methods as follows:
```
private RemoveRedundantLogicalLocationProperties()
{
    foreach (KeyValuePair<string, LogicalLocation logicalLocation> entry in _currentRun.LogicalLocation)
    {
        RemoveRedundantProperties(entry.Key, entry.Value);
    }
}

private void RemoveRedundantProperties(string key, LogicalLocation logicalLocation)
{
    ...
}
```

Reply:
Mostly it's a naming thing. I didn't want to have to read further down the page to know what was being "fixed up". Then once I knew, the renaming and refactoring suggested itself.

Reply:
Done.